scripts/ice-road-pipeline.py

```diff
@@ -124,15 +164,170 @@
+        
+        # Read in cal data and estimate factor
+        df = pd.read_csv(output_csv)
+        median_rfl = df['rfl'].median()
+        road_cal_factor = known_rfl / median_rfl
+        road_cal_factor = str(road_cal_factor) #convert to str for R
+
+        # For each file in <in-dir> 
+        for f in os.listdir(in_dir):
+             
+            # Set names
+            base_las = os.path.basename(f)
+            las_name = os.path.splitext(base_las)[0]
+
+            # R is making a .DS_Store in in_dir, and I can't figure out how or where...
+            # So this will be just a quick check to continue to next iter if encountered.
+            if las_name == ".DS_Store" or las_name == "ice-road":
+                continue 
+
+            rfl_fp = f'{ssa_dir}/{las_name}-rfl.las'
+
+            # Getting a translated "LAS" file
+            # "LAS" in quotations bc I am hiding the rfl here in "Z"
+            # with the intention to do fast IDW in the next step.
+            if os.path.exists(rfl_fp):
+                pass
+            else:
+                subprocess.call(["Rscript", 
+                                f"{scripts_dir}/las_ssa_prep.r", 
+                                f, crs, ni_fp, nj_fp, nk_fp, rfl_fp,
+                                n_e_d_shift,
+                                road_cal_factor,
+                                imu_data])
+
+        
+        rfl_fp = f'{ssa_dir}/rfl-merged.las'
+        rfl_fp_grid = f'{ssa_dir}/rfl-merged.tif'
+
+        # Prep PDAL commands
+        #   ZACH: resolution is 1.0 in laz2dem.py??
+        json_path = join(json_dir, 'combine-rfl.json')
+        json_pipeline = {
+            "pipeline": [
+                rfl_fp,
+                {
+                    "type":"writers.gdal",
+                    "filename":rfl_fp_grid,
+                    "resolution":pix_size, 
+                    "output_type":"idw"
+                }
+            ]
+        }
+        with open(json_path,'w') as outfile:
+            json.dump(json_pipeline, outfile, indent = 2) 
+
+        # Run PDAL commands
+        inp_str = ' '.join(glob(join(ssa_dir, '*.las')))
+        if os.path.exists(rfl_fp_grid):
+            pass    
+        else:
+            cl_call(f'pdal merge {inp_str} {rfl_fp}', log)       
+            cl_call(f'pdal pipeline {json_path}', log)     
+
+        # Prepare inputs needed for SSA raster (vectorized operation)
+        # theta_grid = 180 (perfect backscatter relative to sensor)
+        ssa_fp = f'{ssa_dir}/ssa.tif'
+        rfl_grid = rio.open_rasterio(rfl_fp_grid, masked=True)
+        rfl_grid = rfl_grid.rio.reproject_match(snowoff)
+        ssa_grid = rfl_grid.copy()
+        theta = 180 # based on data it is almost always 179-180
+        cosi=1 # because the reflectance is normalized by incidence angle in las_ssa_prep.R
+
+        # Apply ART - assuming all pixels are snow, directly solve SSA (pretty speedy)
+        k_ice = 1.8983928418833426e-06
```

Comment:
Can you add comments above these so other folks know what these values are? And if there are any that we might want to play with someday?

Reply:
Certainly. I think this only impacts the grain shape parameters, but yeah tuning these values is crucial. The model is quite sensitive to these values. I will incorporate the grain shape as parameters into the new function that can be adjusted if needed.

```python
# The cosine of light incidence angle with surface normal.
# Because the reflectance is normalized by incidence angle in las_ssa_prep.R, 
# cosi is fixed at 1.
cosi=1

# Apply AART - assuming all pixels are snow, directly solve SSA
# This is the imaginary refractive index of ice for the wavelength at 1064 nm (fixed). 
k_ice = 1.8983928418833426e-06

# This is the wavelength in nanometers, then multiplied to meters for computation (fixed).
wl = 1064*1e-9

# This is the density of ice in kg/m3 (fixed).
d_ice = 917

# These are grain shape coefficients that can be tuned based on the expected shape (tunable).
# g=0.85 and b=1.6 are from Libois et al. (2013), however, other shapes may be used to match conditions.
g = 0.85
b = 1.6
```